### PR TITLE
Flog

### PR DESCRIPTION
**Pull request by rafamanzo**
That's one small step for metric collection, a giant leap for Mezuro

### CODE REVIEW

**Comment by diegoamc** [comment]
Haha!

2015-03-30 16:03 GMT-03:00 Rafael Reggiani Manzo <notifications@github.com>:

> It's a small step metric collection, but a big step for the Mezuro Project.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mezuro/kalibro_processor/pull/84
> Commit Summary
>
>    - MetricFu runner class that makes the call on terminal
>    - Added collect_metrics method to Metric Fu collector class.
>    - Implemented metric fu flog parser. Now we need to use it in
>    collector.
>    - Added some tests. Still missing some for Flog class.
>    - Unit tests for metric_fu parser.
>    - Adding processing to the parser methods to create module result
>    - Added test for Flog parser, which is now producing a module result
>    - Flog parser is now creating metric results
>    - Added metric descriptions and moved collectors_descriptions
>    - Started separation of acceptance tests for analizo and metric_fu
>    - [WIP] Working on the acceptance test for flog parser
>    - [WIP] Acceptance test for flog parser, waiting update on kalibro java
>    - Partial fixes after rebase from dev
>    - Gemfile.lock updated after rebase
>    - rspec-rails updated in order to get the missing gems after rebase
>    - Pending unit tests passing
>    - Adapted MetricFu collector and Flog parser to the latest
>    modifications
>    - Fixes MetricCollector finding endpoint and lists just supported
>    metrics
>    - Fixed wrong granularity for Flog metric
>    - Flog does not add empty MetricResults
>
> File Changes
>
>    - *M* Gemfile
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-0> (3)
>    - *M* Gemfile.lock
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-1>
>    (104)
>    - *M* app/controllers/metric_collectors_controller.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-2> (3)
>    - *D* config/collectors_descriptions.yml
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-3> (4)
>    - *M* features/runner.feature
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-4>
>    (11)
>    - *M* features/step_definitions/runner_steps.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-5>
>    (23)
>    - *M* features/support/env.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-6> (2)
>    - *M* lib/metric_collector/base.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-7> (2)
>    - *M* lib/metric_collector/native.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-8> (6)
>    - *M* lib/metric_collector/native/analizo.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-9> (2)
>    - *A* lib/metric_collector/native/descriptions.yml
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-10>
>    (9)
>    - *A* lib/metric_collector/native/metric_fu.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-11>
>    (10)
>    - *A* lib/metric_collector/native/metric_fu/collector.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-12>
>    (34)
>    - *A* lib/metric_collector/native/metric_fu/metrics.yml
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-13>
>    (21)
>    - *A* lib/metric_collector/native/metric_fu/parser.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-14>
>    (21)
>    - *A* lib/metric_collector/native/metric_fu/parser/flog.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-15>
>    (58)
>    - *A* lib/metric_collector/native/metric_fu/parser/interface.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-16>
>    (11)
>    - *A* lib/metric_collector/native/metric_fu/runner.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-17>
>    (36)
>    - *M* spec/controllers/metric_collectors_controller_spec.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-18>
>    (4)
>    - *M* spec/factories/granularities.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-19>
>    (7)
>    - *M* spec/factories/kalibro_configurations.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-20>
>    (6)
>    - *M* spec/factories/kalibro_modules.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-21>
>    (11)
>    - *M* spec/factories/metric_configurations.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-22>
>    (15)
>    - *A* spec/factories/metric_fu_metric_collector_lists.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-23>
>    (11)
>    - *A* spec/factories/metric_fu_runners.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-24>
>    (9)
>    - *M* spec/factories/metrics.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-25>
>    (19)
>    - *M* spec/factories/repositories.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-26>
>    (8)
>    - *A* spec/lib/metric_collector/native/metric_fu/collector_spec.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-27>
>    (24)
>    - *A* spec/lib/metric_collector/native/metric_fu/parser/flog_spec.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-28>
>    (30)
>    - *A* spec/lib/metric_collector/native/metric_fu/parser_spec.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-29>
>    (20)
>    - *A* spec/lib/metric_collector/native/metric_fu/runner_spec.rb
>    <https://github.com/mezuro/kalibro_processor/pull/84/files#diff-30>
>    (43)
>
> Patch Links:
>
>    - https://github.com/mezuro/kalibro_processor/pull/84.patch
>    - https://github.com/mezuro/kalibro_processor/pull/84.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mezuro/kalibro_processor/pull/84>.
>



-- 
Diego.

File: lib/metric_collector/native.rb [hunk 1]
```diff
@@ -1,8 +1,10 @@
 require 'metric_collector/native/analizo'
+require 'metric_collector/native/metric_fu'
 
 module MetricCollector
   module Native
-    ALL = {"Analizo" => MetricCollector::Native::Analizo}
+    #TODO Add metric_fu to the list of native metric collectors
```
Comment (rafamanzo) [inline lib/metric_collector/native.rb]: This TODO is done

File: lib/metric_collector/native/metric_fu/collector.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    module MetricFu
+      class Collector < MetricCollector::Base
+        def initialize
+          description = YAML.load_file("#{Rails.root}/lib/metric_collector/native/descriptions.yml")["metric_fu"]
+          super("MetricFu", description, parse_supported_metrics)
+        end
+
+        def collect_metrics(code_directory, wanted_metric_configurations, processing)
+          self.wanted_metrics = wanted_metric_configurations
+          runner = Runner.new(repository_path: code_directory)
```
Comment (diegoamc) [inline lib/metric_collector/native/metric_fu/collector.rb]: Can't we think of a better name for the Runner class? When I first read it, I didn't know what it was about.

Comment (pedroscocco) [inline lib/metric_collector/native/metric_fu/collector.rb]: How about `Analyzer`? We could keep the method name `run`, becoming `analyzer.run`

Comment (diegoamc) [inline lib/metric_collector/native/metric_fu/collector.rb]: I think it's better than Runner. (:

2015-03-31 0:27 GMT-03:00 pedroscocco <notifications@github.com>:

> In lib/metric_collector/native/metric_fu/collector.rb
> <https://github.com/mezuro/kalibro_processor/pull/84#discussion_r27450673>
> :
>
> > @@ -0,0 +1,34 @@
> > +module MetricCollector
> > +  module Native
> > +    module MetricFu
> > +      class Collector < MetricCollector::Base
> > +        def initialize
> > +          description = YAML.load_file("#{Rails.root}/lib/metric_collector/native/descriptions.yml")["metric_fu"]
> > +          super("MetricFu", description, parse_supported_metrics)
> > +        end
> > +
> > +        def collect_metrics(code_directory, wanted_metric_configurations, processing)
> > +          self.wanted_metrics = wanted_metric_configurations
> > +          runner = Runner.new(repository_path: code_directory)
>
> How about Analyzer? We could keep the method name run, becoming
> analyzer.run
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mezuro/kalibro_processor/pull/84/files#r27450673>.
>



-- 
Diego.

File: lib/metric_collector/native/metric_fu/parser/flog.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                  end
+                end
+              end
+            end
+          end
+
+          private
+
+          def self.module_name_prefix(file_name)
+            without_extension = file_name.split('.')[0].to_s
```
Comment (pedroscocco) [inline lib/metric_collector/native/metric_fu/parser/flog.rb]: Isn't file_name already a string? `to_s`is unnecessary.

File: lib/metric_collector/native/metric_fu/parser/flog.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            without_extension.gsub('/', '.')
+          end
+
+          def self.module_name_sufix(module_name)
+            sufix = module_name.gsub(/(#)|(::)/, ".").split('.')[-1]
```
Comment (pedroscocco) [inline lib/metric_collector/native/metric_fu/parser/flog.rb]: Should split directly with the regex (which also has unnecessary parenthesis).

File: lib/metric_collector/native/metric_fu/parser/flog.rb [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    module MetricFu
+      module Parser
+        class Flog < MetricCollector::Native::MetricFu::Parser::Interface
+          def self.parse(flog_output, processing, metric_configuration)
+            flog_output[:method_containers].each do |method_container|
+              unless method_container[:path].blank?
+                name_prefix = module_name_prefix(method_container[:path])
+                method_container[:methods].each do |name, result|
+                  value = result[:score]
+                  name_sufix = module_name_sufix(name)
```
Comment (pedroscocco) [inline lib/metric_collector/native/metric_fu/parser/flog.rb]: Sufix => Suffix

File: lib/metric_collector/native/metric_fu/parser/flog.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
+module MetricCollector
+  module Native
+    module MetricFu
+      module Parser
+        class Flog < MetricCollector::Native::MetricFu::Parser::Interface
+          def self.parse(flog_output, processing, metric_configuration)
```
Comment (pedroscocco) [inline lib/metric_collector/native/metric_fu/parser/flog.rb]: While strange, it is possible that there are multiple classes in the same file that have methods with the same name. Discarding the module and class names can possibly cause ambiguities, and maybe isn`t a good idea.

File: lib/metric_collector/native/metric_fu/parser/flog.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+          end
+
+          private
+
+          def self.module_name_prefix(file_name)
+            without_extension = file_name.split('.')[0].to_s
+            without_extension.gsub('/', '.')
+          end
+
+          def self.module_name_sufix(module_name)
```
Comment (pedroscocco) [inline lib/metric_collector/native/metric_fu/parser/flog.rb]: While strange, it is possible that there are multiple classes in the same file that have methods with the same name. Discarding the module and class names can possibly cause ambiguities, and maybe isn`t a good idea.

File: lib/metric_collector/native/metric_fu/parser/flog.rb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+          def self.parse(flog_output, processing, metric_configuration)
+            flog_output[:method_containers].each do |method_container|
+              unless method_container[:path].blank?
+                name_prefix = module_name_prefix(method_container[:path])
+                method_container[:methods].each do |name, result|
+                  value = result[:score]
+                  name_sufix = module_name_sufix(name)
+                  unless name_sufix.blank?
+                    module_name = name_prefix + name_sufix
+                    granularity = (module_name == name_prefix ? Granularity::CLASS : Granularity::METHOD)
```
Comment (pedroscocco) [inline lib/metric_collector/native/metric_fu/parser/flog.rb]: `module_name == name_prefix` is never true

File: spec/lib/metric_collector/native/metric_fu/runner_spec.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
+require 'rails_helper'
+require 'metric_collector'
+
+describe MetricCollector::Native::MetricFu::Runner, :type => :model do
+  let!(:repository_path) { Dir.pwd }
+  subject { MetricCollector::Native::MetricFu::Runner.new(repository_path: repository_path) }
+
+  describe 'initialize' do
+    it 'is expected to have a yaml_path' do
+      expect(MetricCollector::Native::MetricFu::Runner.new(repository_path: repository_path).yaml_path).to_not be_nil
```
Comment (pedroscocco) [inline spec/lib/metric_collector/native/metric_fu/runner_spec.rb]: Could use subject here

**Comment by danielkza** [comment]
It seems Travis is having some issues with our test repositories. Is Gitorious going down? Maybe we could set up a Github mirror temporarily.

**Comment by mezurometrics** [comment]
We for sure need to make our tests independent of gitorious.

Em qua, 1 de abr de 2015 às 22:08, Daniel Miranda <notifications@github.com>
escreveu:

**Pull request merged**